### DISCUSSION TITLE
3.7.3-DEV on 5.2.1?

### DISCUSSION THREAD

**Discussion by St-Aris** [fieldOfView/Cura-OctoPrintPlugin#299] [General]
Tried installing this to resolve memory leak issues - but it does not install.  Should it work on 5.2.1 - installer I used was listed as 5.0.

Thanks,

**Comment by fieldOfView** [comment]
How did you try installing it, and can you elaborate on "it does not install"?
The 5.0 curapackage should work on Cura 5.0 and newer.

**Reply by St-Aris** [reply]
Thank you for the reply!  Apologies for the brevity previously - I was not sure if i'd get a reply :)

I took the .curapackage, and dropped it onto cura.  It said it would install on restart - so I restarted.
Upon restarting - it says "The plugin OctoPrintPlugin could not loaded.  Reinstalling the plugin might solve the issue". The message included a button which offered to remove the plugin.

**Reply by fieldOfView** [reply]
Thanks for the elaboration.

Could you please post a link to your cura.log file? It can be found via Help -> Show configuration folder.
Sending attachments via email does not work on github; you have to upload your Cura.log file somewhere (google drive, dropbox, yousendit) and post a link here.

**Reply by St-Aris** [reply]
Hi.
Here is the log.  I cleared it, then started cura, installed the plugin, and then restarted cura.

https://mega.nz/file/EUBUCC6a#OY6HYB-FnKg_zfuVc1NLWXEqH5ef0vESfemqlxQQJ0w

**Reply by fieldOfView** [reply]
Thanks for the log, it was helpful in fixing an error.

I have recreated the 3.7.3-dev release, which now also clarifies that the 5.0 version is for 5.0 - current versions.
https://github.com/fieldOfView/Cura-OctoPrintPlugin/releases/tag/v3.7.3-DEV
Please let me know if this works. I would also be interested to know if it fixes any memory-leak you had.

**Reply by St-Aris** [reply]
Install went fine - thank you!  I'll leave Cura running for 24 hours and let you know about any memory leak issues.

**Reply by St-Aris** [reply]
Unfortunately memory use has gone up overnight.  Was between 600-700 MB yesterday - now sitting at about 1.2GB.  I can't be certain that it is your plugin - perhaps i'll uninstall  it and repeat the experiment and report back.

**Reply by fieldOfView** [reply]
Can you give me another log, to see if the newer version of zeroconf is being used?
It could help troubleshooting to disable the USB Printing and Ultimaker Network Connection plugins.

**Reply by St-Aris** [reply]
Will this snippet suffice?

2022-11-16 12:23:01,043 - DEBUG - [MainThread] OctoPrintPlugin.OctoPrintOutputDevicePlugin.<module> [50]: The zeroconf module is already imported; trying to flush it so we can import our own version
2022-11-16 12:23:01,289 - DEBUG - [MainThread] OctoPrintPlugin.OctoPrintOutputDevicePlugin.<module> [83]: Using included Zeroconf module version 0.39.1
2022-11-16 12:23:01,325 - INFO - [MainThread] UM.PluginRegistry.loadPlugin [493]: Loaded plugin OctoPrintPlugin 3.7.3-DEV

**Reply by St-Aris** [reply]
I had the UM3 plugin disabled already - but i've now disabled the USB one too.

**Reply by St-Aris** [reply]
So, overnight, sitting idle, Cura is at 2GB.  I'm going to disable octoprint plugin, and see if that makes any difference. 

Edit:  Removed Octorpint plugin - sitting at 386MB used.  Lets see what happens.

**Reply by fieldOfView** [reply]
Thanks very much for doing these tests.
I assume it is sitting idle at the "Prepare" screen, not the "Monitor" screen?

**Reply by St-Aris** [reply]
Yes, prepare/monitor screen (I think).

**Reply by fieldOfView** [reply]
Erm, which one? If it is the monitor-screen, then the culprit could be eg the webcam-display.

**Reply by St-Aris** [reply]
Sorry - I meant prepare/preview - though I cannot be 100% certain - I wasn't paying attention.  I'll pay more attention during my future tests.

I had webcam functionality switched off from within the octoprint cura plugin settings.

**Reply by St-Aris** [reply]
No memory leak with the plugin uninstalled.  525MB this morning - and it was that yesterday afternoon.  I imported a large-ish STL, sliced it, and left it on the prepare screen.

I'll install the DEV plugin, and do the same (stay in prepare), and let you know.

**Reply by St-Aris** [reply]
Interesting - it was about 420MB at startup.  Loaded an STL and sliced it - went up to 435.  Went to preview screen - went up to 535.  Looks like when you go to preview, it loads something up to process that part of the code (there was a short pause before preview started).  I went to Monitor just for completeness, and will leave it on the Prepare screen  for 24 hours.

**Reply by fieldOfView** [reply]
One more thing to test is to uncheck the "Automatically discover local OctoPrint instances" box in the Connect to OctoPrint dialog.

**Reply by fieldOfView** [reply]
>  Went to preview screen - went up to 535. 

Displaying the layerdata takes memory. It is not a matter of handing the G-Code to the GPU and being done with it ;-)

**Reply by St-Aris** [reply]
Ok, unchecked that - and restarted for good measure.  Starting memory footprint looks the same - we'll see tomorrow.
Is the zeroconf version I posted above the one you were expecting?

**Reply by fieldOfView** [reply]
Yes, that is the latest and greatest, as distributed with the plugin. If the memory footprint remains to grow with the disciver feature turned off (ie: without using zeroconf), then I'm probably barking up the wrong tree.

**Reply by St-Aris** [reply]
Interesting.  When I first looked at it this morning - it was sitting at 1.5GB - so yes appears to be a leak - but was less of a leak than my last test which was more like 2GB - but that could well be spurious.  It was sitting in the prepare screen this time.

Then I did some work on the PC and tried to use Cura (without restarting it) - it was hesitant and unresponsive for a short time, but them came to life.  I later looked at the memory usage in the task manager - and back to about 560MB.  Not sure what happened there.    May have to test again as I did a few other things on the PC at the time too.

Automatic discover was off - but I am referring to the Octoprint instance as octopi.local instead of the IP address - not sure if that would make a difference as I presume it is using the underlying Windows 10 bonjour/zeroconf resolver.

**Reply by St-Aris** [reply]
Ok some progress.  This morning it was sitting at 1.4GB - a leak.  Tried all sorts of things like loading and slicing some STL's - made no difference.  Then I went into the Configure Octoprint dialogue - Preferences -> Configure Cura -> Printers -> Connect Octoprint.  The leaked memory seems to have been released and it went back down to about 560MB.  Not sure what happens behind the scenes when you go into this dialogue - but perhaps this is a clue for you.

**Reply by fieldOfView** [reply]
> Then I went into the Configure Octoprint dialogue - Preferences -> Configure Cura -> Printers -> Connect Octoprint. 

Did you do anything while there? Or just open the dialog?

**Reply by St-Aris** [reply]
Just opened the dialog.  Does the plugin perhaps try to automatically reconnect when you do that?

**Reply by St-Aris** [reply]
I wasn't exactly testing for this specifically - just messing around and noticed. I will try this again tomorrow and look more closely.

When I go into the dialog it does say 'Checking the API key' - so it must be doing something.  Having restarted Cura - it falls back to about 450MB memory use with a model loaded in and sliced.  When I go to preview, it jumps to about 530MB - and if I go to monitor, about 536MB.

Are you unable to reproduce this yourself?

**Reply by fieldOfView** [reply]
Opening the dialog will destroy and recreate the zeroconf object if discovery is enabled, but it should not exist in the first place with the "Automatically discover local OctoPrint instances" box unchecked.

> Are you unable to reproduce this yourself?

In all honesty, I never leave Cura open long enough to notice. In addition, I don't use printers connected via OctoPrint myself much lately; I hope that will change again in the future.

**Reply by fieldOfView** [reply]
The last time I investigated a memory leak that was not caused by zeroconf, it turned out the memory was leaking somewhere in Qt. This makes things hard to debug:
https://github.com/fieldOfView/Cura-OctoPrintPlugin/issues/88#issuecomment-431784300

What *could* be happening is that there's a leak in the network calls I make periodically to get the state of OctoPrint. 

I'll try to make some time to see if I can reproduce the leak.

**Reply by St-Aris** [reply]
Let me know if there are any further tests you may find helpful for me to try.

Microsoft SDK provides a utility called UMDH - I could potentially figure it out and capture some memory usage info - but I suspect it might be better if someone familiar with the underlying code did that.

**Reply by St-Aris** [reply]
Playing around some more.  Longer you leave it, the more it leaks.  Sometimes it releases the memory - sometimes not (I have to restart to release memory).

**Comment by St-Aris** [comment]
Hope you are still interesting in what i'm seeing - if not, let me know.

Left it on for several days - it went to about 3GB.  I went to the octoprint menu and clicked on 'connect to octoprint' and it suddenly cleared the memory down to 500mb.  I did this again several days later when it went to about 2.2gb - and I was unable to clear the excess memory like this again.  What I did see though was that while I could print from cura directly to octoprint - the monitor page on cura did not show any live information about the print as it usually does.  I restarted, and then this worked again.  I can't quite pinpoint what is going on and reliably reproduce any of these scenarios - it is just general weirdness.  Hope this helps.

**Comment by St-Aris** [comment]
I note there is a new release which fixes some memory leaks - this particular memory leak is still present though.

**Comment by fieldOfView** [comment]
Good to know. The recent update was mostly the same as the 3.7.3-DEV you tested.
I have not had much time to further investigate this, but it is on my todo-list.

**Comment by St-Aris** [comment]
Great - thanks for the update.  If you need any further info or repro/debug info - just comment on this thread and i'll do my best to help.
